### PR TITLE
Common - Filter for ACRE radio IDs when serializing objects

### LINKED ISSUE #636: Deep copy/paste will include ACRE radio ID's
**Arma 3 Version:** `2.04`
**CBA Version:** `3.15.5` (stable / dev + commit hash)
**ZEN Version:** `1.11.1` (stable / dev + commit hash)

**Mods:**
```
- CBA_A3
- ZEN
- ACE
- ACRE
```

**Description:**
When using deep copy/paste on units with ACRE radios, their radio ID's are copied as well, creating two sets of the same radio and has the potential to completely break both the original and deep-copied radios.

**Expected behavior:**
Each unit should have their own radios with their own radio ID's.

**Steps to reproduce:**
1. Equip a loadout on yourself with ACRE radios
2. Deep copy yourself
3. Deep paste yourself
4. Notice that executing `systemChat str (["ACRE_PRC152", _this] call acre_api_fnc_getAllRadiosByType)` in the ZEN init box for both units will return the same radios.

**Where did the issue occur?**
- Dedicated / Self-Hosted Multiplayer

**Additional information:**
ACRE provides the function `acre_api_fnc_filterUnitLoadout` which will filter a unit loadout array to ensure this issue doesn't happen. Most likely a CfgPatches check for ACRE + using this function in `zen_common_fnc_serializeObjects` would fix the issue and I might slap the PR together myself if I get time.

### PR DESCRIPTION
**Pull request by mjc4wilton**
**When merged this pull request will:**
- Add CfgPatches check for ACRE via Common `postInit.sqf`
- If ACRE is installed, pass the loadout through to `acre_api_fnc_filterUnitLoadout` to remove replace ACRE radio IDs with base radios
- Fix #636

### CODE REVIEW

File: addons/common/XEH_postInit.sqf [hunk 1]
```diff
@@ -399,3 +399,9 @@ if (isServer) then {
     [QGVAR(createZeus), LINKFUNC(createZeus)] call CBA_fnc_addEventHandler;
     [QGVAR(deserializeObjects), LINKFUNC(deserializeObjects)] call CBA_fnc_addEventHandler;
 };
+
+// Check for installed mods
+GVAR(hasACRE) = false;
+if (isClass(configFile >> "CfgPatches" >> "acre_main")) then {
+    GVAR(hasACRE) = true;
+};
```
Comment (mharis001) [inline addons/common/XEH_postInit.sqf:407]: Doesn't need the `if` block.

File: addons/context_actions/CfgContext.hpp [hunk 2]
```diff
@@ -236,7 +236,7 @@ class EGVAR(context_menu,actions) {
         };
         class Copy {
             displayName = "$STR_3DEN_Display3DEN_MenuBar_EntityCopy_text";
-            statement = QUOTE(GVAR(loadout) = getUnitLoadout _hoveredEntity);
+            statement = QUOTE(if !(GVAR(hasACRE)) then {GVAR(loadout) = getUnitLoadout _hoveredEntity;} else {GVAR(loadout) = [getUnitLoadout _hoveredEntity] call acre_api_fnc_filterUnitLoadout;});
```
Comment (mharis001) [inline addons/context_actions/CfgContext.hpp:239]: This seems weird, also `GVAR(hasACRE)` is incorrect here.

**Review by mharis001 — CHANGES_REQUESTED** [review]
If we go with this approach then I would want to add TFAR support (doesn't need to be in this PR). We could also try and hold out for the extended loadout framework proposed for CBA but not sure when (or if) that will be merged.

Overall, seems okay, but maybe we should have our own get/set loadout functions in `common` instead of duplicating code.

**Comment by Kexanone** [comment]
@mharis001
When do we prefer event handlers over functions?

Also I'm not really sure if the `safe` postfix is really necessary.

File: addons/context_actions/CfgContext.hpp [hunk 3]
```diff
@@ -236,7 +236,7 @@ class EGVAR(context_menu,actions) {
         };
         class Copy {
             displayName = "$STR_3DEN_Display3DEN_MenuBar_EntityCopy_text";
-            statement = QUOTE(GVAR(loadout) = getUnitLoadout _hoveredEntity);
+            statement = QUOTE(GVAR(loadout) = [ARR_1(_hoveredEntity)] call EFUNC(common,getUnitLoadoutSafe););
```
Comment (Kexanone) [inline addons/context_actions/CfgContext.hpp:239]:
```suggestion
            statement = QUOTE(GVAR(loadout) = _hoveredEntity call EFUNC(common,getUnitLoadoutSafe););
```

File: addons/common/functions/fnc_getUnitLoadoutSafe.sqf [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+params ["_unit"];
+
+private _loadout = getUnitLoadout _unit;
+
+// ACRE Sanitization
+if (isClass(configFile >> "CfgPatches" >> "acre_main")) then {
+    _loadout = [_loadout] call acre_api_fnc_filterUnitLoadout;
+}
+
+return _loadout;
```
Comment (Kexanone) [inline addons/common/functions/fnc_getUnitLoadoutSafe.sqf:27]:
```suggestion
_loadout // Return value
```
This is not `C++` 😅

**Comment by mjc4wilton** [comment]
> Also I'm not really sure if the `safe` postfix is really necessary.

Did it so its not just getUnitLoadout wrapped in a macro. Not necessary but helps maintain syntax highlighting

File: addons/common/XEH_PREP.hpp [hunk 5]
```diff
@@ -28,6 +28,7 @@ PREP(getPylonTurret);
 PREP(getSelectedUnits);
 PREP(getSelectedVehicles);
 PREP(getSideIcon);
+PREP(getUnitLoadoutSafe);
```
Comment (mharis001) [inline addons/common/XEH_PREP.hpp:31]: Change function name to `getUnitLoadout`.

File: addons/common/functions/fnc_getUnitLoadoutSafe.sqf [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ * Public: No
+ */
+
+params ["_unit"];
+
+private _loadout = getUnitLoadout _unit;
+
+// ACRE Sanitization
+if (isClass(configFile >> "CfgPatches" >> "acre_main")) then {
```
Comment (mharis001) [inline addons/common/functions/fnc_getUnitLoadoutSafe.sqf:23]:
```suggestion
if (isClass (configFile >> "CfgPatches" >> "acre_main")) then {
```

File: addons/common/functions/fnc_getUnitLoadoutSafe.sqf [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+ * Returns the unit loadout array (sanitized)
+ *
+ * Arguments:
+ * 0: Unit <OBJECT>
+ *
+ * Return Value:
+ * Unit Loadout Array <ARRAY>
+ *
+ * Example:
+ * [player] call zen_common_fnc_getUnitLoadoutSafe
```
Comment (mharis001) [inline addons/common/functions/fnc_getUnitLoadoutSafe.sqf:13]:
```suggestion
 * [_unit] call zen_common_fnc_getUnitLoadout
```

File: addons/common/functions/fnc_getUnitLoadoutSafe.sqf [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+params ["_unit"];
+
+private _loadout = getUnitLoadout _unit;
+
+// ACRE Sanitization
+if (isClass(configFile >> "CfgPatches" >> "acre_main")) then {
+    _loadout = [_loadout] call acre_api_fnc_filterUnitLoadout;
+};
+
+_loadout // Return value
```
Comment (mharis001) [inline addons/common/functions/fnc_getUnitLoadoutSafe.sqf:27]:
```suggestion
_loadout
```

File: addons/context_actions/CfgContext.hpp [hunk 9]
```diff
@@ -236,7 +236,7 @@ class EGVAR(context_menu,actions) {
         };
         class Copy {
             displayName = "$STR_3DEN_Display3DEN_MenuBar_EntityCopy_text";
-            statement = QUOTE(GVAR(loadout) = getUnitLoadout _hoveredEntity);
+            statement = QUOTE(GVAR(loadout) = _hoveredEntity call EFUNC(common,getUnitLoadoutSafe););
```
Comment (mharis001) [inline addons/context_actions/CfgContext.hpp:239]:
```suggestion
            statement = QUOTE(GVAR(loadout) = _hoveredEntity call EFUNC(common,getUnitLoadout));
```

File: addons/common/functions/fnc_getUnitLoadoutSafe.sqf [hunk 10]
```diff
@@ -0,0 +1,27 @@
+#include "script_component.hpp"
+/*
+ * Author: mjc4wilton
+ * Returns the unit loadout array (sanitized)
```
Comment (mharis001) [inline addons/common/functions/fnc_getUnitLoadoutSafe.sqf:4]:
```suggestion
 * Returns the given unit's loadout. Handles filtering items with unique radio IDs.
```

**Review by mharis001 — COMMENTED** [review]
Looks good. Just some cleanup.

File: addons/common/functions/fnc_exportMissionSQF.sqf [hunk 11]
```diff
@@ -115,7 +115,8 @@ private _fnc_processInventory = {
         _outputObjects pushBack ["['%1', 'onEachFrame', {", _nextFrameHandle];
         _outputObjects pushBack "    params [""_unit""];";
         if !(_object call FUNC(hasDefaultInventory)) then {
-            _outputObjects pushBack ["    _unit setUnitLoadout %1;", getUnitLoadout _object];
+            private _loadout = [_object] call FUNC(getUnitLoadoutSafe);
```
Comment (mharis001) [inline addons/common/functions/fnc_exportMissionSQF.sqf:118]: Function name needs to be updated here and in other places.

**Review by mharis001 — APPROVED** [review]

**Review by Kexanone — APPROVED** [review]

**Pull request merged**